File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/Common/DefaultFont.java
package com.spvessel.spacevil.Common;

import com.spvessel.spacevil.Flags.EmbeddedFont;

import java.awt.Font;
import java.io.InputStream;

final class DefaultFont {
    private static DefaultFont _instance;

    private DefaultFont() {}

    public static DefaultFont getInstance() {
        if (_instance == null)
            _instance = new DefaultFont();
        return _instance;
    }

    private InputStream privateFontCollection = null;
    private Font _default_font = null;

    public void setDefaultFont(Font font) {
        _default_font = font;
    }

    public void setDefaultFont(InputStream font_stream) {
        try {
            privateFontCollection = font_stream;
            Font font = Font.createFont(Font.TRUETYPE_FONT, privateFontCollection);
            _default_font = font.deriveFont(Font.PLAIN, 12);
        } catch (Exception ex) {

        }
    }

    public Font getDefaultFont() {
        if (_default_font == null) {
            addFontFromMemory();
            try {
                Font font = Font.createFont(Font.TRUETYPE_FONT, privateFontCollection);
                _default_font = font.deriveFont(Font.PLAIN, 12);
            } catch (Exception ex) {

            }
        }
        return _default_font;
    }

    public Font getDefaultFont(int size) {
        // Console.WriteLine(size);
        if (size == 0) {
            size = 10;
        }
        if (_default_font == null) {
            addFontFromMemory();
            try {
                Font font = Font.createFont(Font.TRUETYPE_FONT, privateFontCollection);
                _default_font = font.deriveFont(Font.PLAIN, 12);
            } catch (Exception ex) {

            }
        }
        return _default_font.deriveFont(Font.PLAIN, size);
    }

    public Font getDefaultFont(int style, int size) {
        // Console.WriteLine(size);
        if (size == 0) {
            size = 10;
        }
        if (_default_font == null) {
            addFontFromMemory();
            try {
                Font font = Font.createFont(Font.TRUETYPE_FONT, privateFontCollection);
                _default_font = font.deriveFont(Font.PLAIN, 12);
            } catch (Exception ex) {

            }
        }
        return _default_font.deriveFont(style, size);
    }

    private void addFontFromMemory() {
        if (_embedded_font == null || _embedded_font == "")
            setDefaultEmbeddedFont(EmbeddedFont.UBUNTU);

        privateFontCollection = DefaultFont.class.getResourceAsStream(_embedded_font);
    }

    private String _embedded_font = null;

    public void setDefaultEmbeddedFont(EmbeddedFont font) {
        switch (font) {
            case UBUNTU:
                _embedded_font = "/fonts/Ubuntu-Regular.ttf";
                break;

            default:
                _embedded_font = "/fonts/Ubuntu-Regular.ttf";
                break;
        }
    }
}

File: CSharp/SpaceVIL/SpaceVIL/Common/DefaultFont.cs
using System;
using System.Drawing;
using System.Drawing.Text;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using SpaceVIL.Core;

namespace SpaceVIL.Common
{
    internal sealed class DefaultFont
    {
        private DefaultFont() { }

        private PrivateFontCollection _privateFontCollection = null;

        private String _embeddedFont = null;

        private Font _defaultFont;

        private const int _defFontSize = 12;
        
        private static DefaultFont _instance;

        internal static DefaultFont GetInstance()
        {
            if (_instance == null)
                _instance = new DefaultFont();
            return _instance;
        }

        internal void SetDefaultFont(Font font)
        {
            _defaultFont = font;
        }

        internal Font GetDefaultFont(int size = _defFontSize)
        {
            if (size == 0)
            {
                size = _defFontSize;
            }
            if (_defaultFont == null)
            {
                AddFontFromMemory();
                _defaultFont = new Font(_privateFontCollection.Families[0], _defFontSize, FontStyle.Regular);
            }
            return GraphicsMathService.ChangeFontSize(size, _defaultFont);
        }
        internal Font GetDefaultFont(FontStyle style, int size)
        {
            if (size == 0)
            {
                size = _defFontSize;
            }
            if (_defaultFont == null)
            {
                AddFontFromMemory();
                _defaultFont = new Font(_privateFontCollection.Families[0], size, style);
            }
            return new Font(_defaultFont.FontFamily, size, style);
        }

        private void AddFontFromMemory()
        {
            if (String.IsNullOrEmpty(_embeddedFont))
                SetDefaultEmbeddedFont(EmbeddedFont.Ubuntu);

            _privateFontCollection = new PrivateFontCollection();

            Stream fontStream = Assembly.GetExecutingAssembly().GetManifestResourceStream(_embeddedFont);
            System.IntPtr data = Marshal.AllocCoTaskMem((int)fontStream.Length);
            byte[] fontdata = new byte[fontStream.Length];
            fontStream.Read(fontdata, 0, (int)fontStream.Length);
            Marshal.Copy(fontdata, 0, data, (int)fontStream.Length);
            _privateFontCollection.AddMemoryFont(data, (int)fontStream.Length);
            fontStream.Close();
            Marshal.FreeCoTaskMem(data);
        }
        
        internal void SetDefaultEmbeddedFont(EmbeddedFont font)
        {
            switch (font)
            {
                case EmbeddedFont.Ubuntu:
                    _embeddedFont = "SpaceVIL.Fonts.Ubuntu-Regular.ttf";
                    break;

                default:
                    _embeddedFont = "SpaceVIL.Fonts.Ubuntu-Regular.ttf";
                    break;
            }
        }
    }
}